editor/plugins/node_3d_editor_plugin.cpp

```diff
@@ -2941,16 +2944,243 @@ void Node3DEditorViewport::_notification(int p_what) {
 
 				geometry->clear_surfaces();
 				geometry->surface_begin(Mesh::PRIMITIVE_LINES);
-				geometry->surface_add_vertex(start_pos);
-				geometry->surface_add_vertex(end_pos);
+
+				Vector3 center = (start_pos + end_pos) / 2;
+				Vector3 camera_dir = (camera->get_transform().origin - center).normalized();
+				real_t offset_distance = 0.01;
+
+				geometry->surface_add_vertex(start_pos + camera_dir * offset_distance);
+				geometry->surface_add_vertex(end_pos + camera_dir * offset_distance);
 				geometry->surface_end();
 
+				geometry_xray->clear_surfaces();
+				geometry_xray->surface_begin(Mesh::PRIMITIVE_LINES);
+				geometry_xray->surface_add_vertex(start_pos);
+				geometry_xray->surface_add_vertex(end_pos);
+				geometry_xray->surface_end();
+
 				float distance = start_pos.distance_to(end_pos);
+				if (distance < 0.001) {
+					distance = 0.0;
+				}
 				ruler_label->set_text(TS->format_number(vformat("%.3f m", distance)));
 
-				Vector3 center = (start_pos + end_pos) / 2;
 				Vector2 screen_position = camera->unproject_position(center) - (ruler_label->get_custom_minimum_size() / 2);
 				ruler_label->set_position(screen_position);
+
+				bool show_components = Input::get_singleton()->is_key_pressed(Key::SHIFT);
+
+				if (show_components) {
+					Ref<ImmediateMesh> triangle_mesh = ruler_triangle_lines->get_mesh();
+					Ref<ImmediateMesh> triangle_mesh_xray = ruler_triangle_lines_xray->get_mesh();
+
+					if (triangle_mesh.is_valid() && triangle_mesh_xray.is_valid()) {
+						Vector3 delta = end_pos - start_pos;
+
+						const real_t threshold = 0.001;
+						if (Math::abs(delta.x) < threshold) {
+							delta.x = 0.0;
+						}
+						if (Math::abs(delta.y) < threshold) {
+							delta.y = 0.0;
+						}
+						if (Math::abs(delta.z) < threshold) {
+							delta.z = 0.0;
+						}
+
+						delta.x = Math::abs(delta.x);
+						delta.y = Math::abs(delta.y);
+						delta.z = Math::abs(delta.z);
```

Comment:
```suggestion
						Vector3 delta = end_pos - start_pos;
						delta = delta.abs();

						const real_t threshold = 0.001;
						if (delta.x < threshold) {
							delta.x = 0.0;
						}
						if (delta.y < threshold) {
							delta.y = 0.0;
						}
						if (delta.z < threshold) {
							delta.z = 0.0;
						}
```
Math::abs should return 0 when passed 0, so I think you can get away with precomputing the absolute values before checking whether they fall under the thresholds. Similar deal goes for lines 3048 through 3064. Still, things are shaping up quite nicely here.

Reply:
Done.